### ISSUE TITLE
SkiaSharp 3.0 support

### ISSUE THREAD

**Issue by maxkatz6** [AvaloniaUI/Avalonia#15503]
### Is your feature request related to a problem? Please describe.

Tracking issue on SkiaSharp 3.0 support status.

#### 11.x Timeline

- [x] [Ensure best-effort compatibility between Avalonia.Skia and SkiaSharp 3.0, while still targeting 2.88.x](https://github.com/AvaloniaUI/Avalonia/pull/14510).
- [x] [Enable Avalonia.Skia with SkiaSharp 3.0 tests](https://github.com/AvaloniaUI/Avalonia/pull/15255).
- [x] [Run Avalonia.Skia with SkiaSharp 3.0 tests on CI.](https://github.com/AvaloniaUI/Avalonia/pull/18981)

#### 12.0 Target

- [x] [Drop support for 2.88.x, only target 3.0](https://github.com/AvaloniaUI/Avalonia/pull/18981)
- [x] [Adjust code base to efficiently use SkiaSharp 3.0 features. Avoid double conversion with SKMatrix44.](https://github.com/AvaloniaUI/Avalonia/pull/18981)
- [x] [Use stable APIs for Metal rendering instead of reflection (only available in SkiaSharp 3.0)](https://github.com/AvaloniaUI/Avalonia/pull/18981)
- [ ] https://github.com/AvaloniaUI/Avalonia/issues/18238

Known issues with SkiaSharp 3.0 on Avalonia:

- [x] Bitmap drawing ignores interpolation mode, as we need to use new API with SKSamplingOptions to support that. `paint.FilterQuality` is a no-op. 
- [x] Developers, who want to try SkiaSharp 3.0, need to manually include NativeAssets packages for their target platforms. It will be automatically solved once we remove 2.88 support. For example, for Linux it would be:
```xml
<PackageReference Include="SkiaSharp.NativeAssets.Linux" Version="3.0.0-preview.3.1" />
```

**Comment by jmacato** [comment]
@cla-avalonia test

**Comment by kerams** [comment]
Android has been working great for me, but I get crashes on startup on iOS. `MetalPlatformGraphics.TryCreate` is failing to dispose of some unmanaged resources by the looks of it.

**Comment by kerams** [comment]
No, I am not particularly experienced in that area. SkiaSharp 3 is also not officially supported in the latest version of Avalonia, so I'm guessing the team would just prefer to wait for the upstream fix anyway.

**Comment by kekekeks** [comment]
Do we aren't allowed to Dispose metal-backed GRContext on iOS, huh

**Comment by jeremy-visionaid** [comment]
The finalizer also crashes - but dispose makes it deterministic. I'm reasonably sure that Skia takes ownership and destroys the command queue when the context is abandoned. So, using GRContext directly then it's also possible to work around with DangerousRetain on the command queue to avoid the double free from the dotnet side, but yeah, it would be better if it got fixed upstream.

**Comment by jeremy-visionaid** [comment]
@maxkatz6 I'm probably a bit late since this already got merged, but did you want to add the DangerousRetain as above? Otherwise it's going to crash on dispose until it gets fixed upstream (though Avalonia will probably get away with it on exit if it's not being disposed deterministically).

https://github.com/mono/SkiaSharp/pull/3258/files#diff-4b5eef515f0228d44e1704e52d67d008d8e8cb75ccf03fac75a4a80b7e39ebd7R95

**Comment by maxkatz6** [comment]
@jeremy-visionaid no, but there was a similar crash fixed by removing old code:
https://github.com/AvaloniaUI/Avalonia/blob/88bfecbee03f1bddbcc193b1890c9ac0665b6b2f/src/iOS/Avalonia.iOS/Metal/MetalPlatformGraphics.cs#L37-L38

This extra GRContext wasn't used for the actual rendering, yet it was crashing iOS/Catalyst/macOS 26. Could it be relevant to your observations?

**Comment by jeremy-visionaid** [comment]
Yeah, here, the command queue is being deterministically disposed after the context, so the GRContext nukes the command queue and then the using statement tries to destroy it a second time which causes a crash. It's a manifestation of the double free I'm referring to. It should also apply to your changes here - it's just not clear if the fault results in an actual failure. So, I think it's likely that Avalonia might just be getting away with it on exit - but in more complex cases it's possible others might observe it non-deterministically from a finalizer thread similarly to how it manifests on MAUI.
